src/base/dof_map_constraints.C

```diff
@@ -1133,32 +1183,80 @@ public:
      * input function \p f gives the arbitrary solution.
      */
 
-    // Loop over all the variables we've been requested to project
-    for (const auto & var : dirichlet.variables)
-      {
-        const Variable & variable = dof_map.variable(var);
+    // Figure out which System the DirichletBoundary objects are
+    // defined for. Verify that all Variables are defined on the same
+    // System.
+    System * system = nullptr;
+    for (const auto & dirichlet : dirichlets)
+      for (const auto & var : dirichlet->variables)
+        {
+          const Variable & variable = dof_map.variable(var);
+          auto current_system = variable.system();
 
-        const FEType & fe_type = variable.type();
+          if (!system)
+            system = current_system;
+          else if (current_system != system)
+            libmesh_error_msg("All variables should be defined on the same System");
+        }
```

Comment:
This part seems a bit costly in your extreme thousand-DirichletBoundary cases.  Sure you don't want it wrapped in ifndef NDEBUG?

Reply:
You're right... this started as a loop with a `break` as soon as a valid System was found but then it went through a couple iterations... I can put it back the way it was originally.